### PR TITLE
docs: introduce "Conversion to Promises" guide

### PR DESCRIPTION
**Pull request by jakovljevic-mladen**
**Description:**
Introduce new guide "Conversion to Promises".

**Related issue (if exists):**
#5927

### CODE REVIEW

**Comment by abarke** [comment]
How would I migrate my code from `toPromise` to `firstValueFrom` when `toPromise` is not available anymore? 

I need the `timeout` functionality to throw an error if the response takes too long and the predicate to filter out any unintended messages.

Here is an example:

```javascript
/**
   * Send a request to the server and expect and response. Reject when an error or timeout occurs.
   * @param channelId
   * @param requestTimeout
   * @private
   */
  private asyncRequest (channelId: string, requestTimeout = 3000): Promise<Message> {
    const request = {
      event: 'request',
      data: channelId,
      id: Date.now()
    }

    // Send request to server
    this.ws$.next(request)

    // Filter out expected response by matching the request and response message id
    const filter = (msg: Message): boolean => msg.id === request.id

    // Create new observable using pipe and convert to a promise
    return this.ws$.pipe(
      first(filter),
      timeout(requestTimeout)
    ).toPromise()
  }
```

**Comment by jakovljevic-mladen** [comment]
> How would I migrate my code from toPromise to firstValueFrom when toPromise is not available anymore?

Hi @abarke, this is a good question and I maybe should have added example of the transition to using new methods instead of just copy/pasting existing examples.

In your case, instead of:

```ts
return this.ws$.pipe(
  first(filter),
  timeout(requestTimeout)
).toPromise()
```

you'd have:

```ts
return lastValueFrom(this.ws$.pipe(
  first(filter),
  timeout(requestTimeout)
))
```

**Comment by jakovljevic-mladen** [comment]
A gentle reminder @cartant about this PR. Ben said this morning on Twitter that the last RC will be the last one before the release, so I would like to know if you have plans to include stuff from this PR to the release. If you do, please let me know so I can rebase as soon as I have time.

**Comment by cartant** [comment]
Sorry. This PR had slipped my mind. I should have reviewed it and included links to the markdown document in https://github.com/ReactiveX/rxjs/pull/6259. If you're able to rebase it, that would be great.

Regarding the release, there is no rush. TBH, the deprecations were not blocking the release. My understanding is the the most important doc-related tasks were to get the breaking changes documented and the v7 docs site done - including supporting the v6 docs on a subdomain. Those tasks are now done, so the release will happen soon. There will be further additions to the docs after the release.

So, yes, we do want to use the work you've done in this PR, but it doesn't have to be completed before the release. Whenever you have time would be great. Thanks.

File: docs_app/content/6-to-7-change-summary.md [hunk 1]
```diff
@@ -163,7 +163,11 @@ This document contains a detailed list of changes between RxJS 6.x and RxJS 7.x,
 
 ### firstValueFrom
 
-- A better, more tree-shakable replacement for `toPromise()` (which is now deprecated). This function allows the user to convert any `Observable` in to a `Promise` that will resolve when the source observable emits its firsr value. If the source observable closes without emitting a value, the returned promise will reject with an `EmptyError`, or it will resolve with a configured `defaultValue`.
+- A better, more tree-shakable replacement for `toPromise()` (which is now deprecated). This function allows the user to convert any `Observable` in to a `Promise` that will resolve when the source observable emits its first value. If the source observable closes without emitting a value, the returned promise will reject with an `EmptyError`, or it will resolve with a configured `defaultValue`.
```
Comment (jakovljevic-mladen) [inline docs_app/content/6-to-7-change-summary.md:166]: > I should have reviewed it and included links to the markdown document in #6259.

@cartant Did you mean to add it to this file? Maybe like this:

```
(which is [now deprecated](/deprecations/to-promise))
```

Comment (cartant) [reply docs_app/content/6-to-7-change-summary.md:166]: I'd add `Details: https://rxjs.dev/deprecations/to-promise` to the `toPromise` deprecations here:

https://github.com/ReactiveX/rxjs/blob/02a90983b1c0965140d49c2b0f999919db16545c/src/internal/Observable.ts#L486-L510

Kinda like this:

https://github.com/ReactiveX/rxjs/blob/02a90983b1c0965140d49c2b0f999919db16545c/src/internal/operators/publish.ts#L12-L18

Comment (jakovljevic-mladen) [reply docs_app/content/6-to-7-change-summary.md:166]: Or like this: https://github.com/ReactiveX/rxjs/pull/5949/files#diff-804122836d26336b7ccd318e42f664919dc07bf973ef16978e0c47753ff113abR510 😁

I'll do it, thanks for suggestion.

File: docs_app/content/6-to-7-change-summary.md [hunk 2]
```diff
@@ -163,7 +163,11 @@ This document contains a detailed list of changes between RxJS 6.x and RxJS 7.x,
...
+- A better, more tree-shakable replacement for `toPromise()` (which is now deprecated). This function allows the user to convert any `Observable` in to a `Promise` that will resolve when the source observable emits its first value. If the source observable closes without emitting a value, the returned promise will reject with an `EmptyError`, or it will resolve with a configured `defaultValue`. Details: https://rxjs.dev/deprecations/to-promise
+
+### lastValueFrom
+
+- A better, more tree-shakable replacement for `toPromise()` (which is now deprecated). This function allows the user to convert any `Observable` in to a `Promise` that will resolve when the source observable emits the last value. If the source observable closes without emitting a value, the returned promise will reject with an `EmptyError`, or it will resolve with a configured `defaultValue`. Details: https://rxjs.dev/deprecations/to-promise
```
Comment (cartant) [inline docs_app/content/6-to-7-change-summary.md:170]: Given that this is Markdown, I think:
```
Details: https://rxjs.dev/deprecations/to-promise
```
Could be:
```
For more information, see the [deprecation guide](https://rxjs.dev/deprecations/to-promise).
```

File: docs_app/content/deprecations/to-promise.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+As a replacement to the deprecated `toPromise()` method, you should use one of the two built in static conversion
+functions {@link firstValueFrom} or {@link lastValueFrom}.
+
+## `lastValueFrom`
+
+The `lastValueFrom` is almost exactly the same as `toPromise()` meaning that it will resolve with the last value that has
+arrived when the Observable completes, but with the difference in behavior when Observable completes without emitting a
+single value. When Observable completes without emitting, `toPromise()` will successfully resolve with `undefined` (thus
+the return type change), while the `lastValueFrom` will reject with the {@link EmptyError}. Thus, the return type of the
```
Comment (cartant) [inline docs_app/content/deprecations/to-promise.md:28]: IDK whether or not the TSDoc-style `{@link whatever}` tags can be used. I would be surprised if the tooling were to turn them into links. Are there examples of this elsewhere? Does it work? In the multicasting docs that I wrote, I just used Markdown links:

```
And [shareReplay](/api/operators/shareReplay) - which is a thin wrapper around the now highly-configurable [share](/api/operators/share) operator.
```

**Review by cartant — COMMENTED** [review]
LGTM. Just a couple for formatting-related issues/questions.

File: docs_app/content/deprecations/to-promise.md (hunk 3 above)
Comment (jakovljevic-mladen) [reply docs_app/content/deprecations/to-promise.md:28]: Actually, they can:

image

Another example:
https://github.com/ReactiveX/rxjs/commit/539b74281eafde149e473f0aeed8a1f75c8ecf9f#diff-859daf5e87a340e04295a22f9ba74ef5a6508fbd7fe99e8fb0a1ecad87b15b3fR156

This is not yet visible at next.rxjs.dev, so here's the excerpt from the local run:

image

Comment (cartant) [reply docs_app/content/deprecations/to-promise.md:28]: TIL. This would have saved me some typing. 😅

**Review by cartant — APPROVED** [review]
LGTM. Thanks.

**Comment by cartant** [comment]
LGTM. I'll leave it for @niklas-wortmann to review and merge.

**Review by niklas-wortmann — APPROVED** [review]
LGTM many thanks for the contribution

**Comment by niklas-wortmann** [comment]
@jakovljevic-mladen could you rebase this so that I can merge it afterwards?

**Pull request merged**